### PR TITLE
feat(smt): add `smt` crate

### PR DESCRIPTION
**Pull request by waddaboo**
## Description

Added the Sparse Merkle Tree (SMT) implementation in Rust.

## Related Issue(s)

Closes: #8 

## Other information
- Please do let me know if anything can be improved.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have run `yarn format` and `yarn compile` without getting any errors
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Additional Questions
- Need to add actual hash functions to test?

### CODE REVIEW

File: crates/smt/src/smt.rs
Comment (sripwoud) [inline crates/smt/src/smt.rs:137]:
```suggestion
        // `zero_node` is saved. This node is used below as the first node (starting from the
```

**Review by sripwoud — APPROVED** [review]
tACK

Thanks @waddaboo 
Nice you added docstring and a README.

I
- checked tests locally
- checked the function signatures and overall code to see if it matches the original TS implementation

I suggested some improvements but they aren't blocking merging this PR imo

File: crates/smt/src/smt.rs
Comment (waddaboo) [reply crates/smt/src/smt.rs:42]: Yes, Rust is returning field `membership` is never read so I added the allow dead code there.

File: crates/smt/examples/smt.rs
Comment (waddaboo) [reply crates/smt/examples/smt.rs:1]: Thanks for the review! 

I saw some other libraries included examples and some did not, so I'm not sure either. Since the README `Usage` section is the same as in `examples`, I think it's ok to remove it?

File: crates/smt/src/smt.rs
Comment (waddaboo) [reply crates/smt/src/smt.rs:1]: Okay, will work on it!

Comment (waddaboo) [reply crates/smt/src/smt.rs:1]: > - derive or implement the `Default`

For this, are you referring to adding a default for the `Node` enum?

Comment (sripwoud) [reply crates/smt/src/smt.rs:1]: I was thinking about implementing it for `SMT`.
But thinking about it again I am not sure that is such a good idea, as it will require to implement `Default` for the underlying `hash` field, and I don't think there is a obvious default hash function candidate we can choose.

So let's forget about it

**Review by 0xjei — APPROVED** [review]
tACK - thank you @waddaboo the code looks polished and precise. I don't have additional comments beside the @sripwoud ones.

File: crates/smt/src/smt.rs
Comment (0xjei) [reply crates/smt/src/smt.rs:38]: Yes, it might make sense to avoid doing these checks at runtime, but 90% of things can be represented by String and BigInt. I don't know if the game is worth the candle. Anyone has any additional comments here?

**Comment by waddaboo** [comment]
@0xjei Thanks for the review! As for the comment on using generics, do I implement it now or wait for further comments?

**Comment by sripwoud** [comment]
> @0xjei Thanks for the review! As for the comment on using generics, do I implement it now or wait for further comments?

No need to implement it now.

Let's merge. @cedoor we need your approval too.
@curryrasul can make post merge comments. If necessary we ll implement other requests for changes later.

Thanks @waddaboo

**Review by cedoor — APPROVED** [review]

**Comment by cedoor** [comment]
@sripwoud thanks. Ready to be merged 👍🏽

**Pull request merged**
